### PR TITLE
Add sst dataset

### PR DESCRIPTION
**Pull request by patpizio**
Related to #1934&mdash;Add the Stanford Sentiment Treebank dataset.

### CODE REVIEW

File: datasets/sst/README.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+[Needs More Information]
+
+### Licensing Information
+
+[Needs More Information]
+
+### Citation Information
+
+[Needs More Information]
```
Comment (lhoestq) [inline datasets/sst/README.md:163]:
````suggestion
```
@inproceedings{socher-etal-2013-recursive,
    title = "Recursive Deep Models for Semantic Compositionality Over a Sentiment Treebank",
    author = "Socher, Richard  and
      Perelygin, Alex  and
      Wu, Jean  and
      Chuang, Jason  and
      Manning, Christopher D.  and
      Ng, Andrew  and
      Potts, Christopher",
    booktitle = "Proceedings of the 2013 Conference on Empirical Methods in Natural Language Processing",
    month = oct,
    year = "2013",
    address = "Seattle, Washington, USA",
    publisher = "Association for Computational Linguistics",
    url = "https://www.aclweb.org/anthology/D13-1170",
    pages = "1631--1642",
}
```
````

File: datasets/sst/README.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+ 'sentence': 'Yet the act is still charming here .',
+ 'tokens': 'Yet|the|act|is|still|charming|here|.',
+ 'tree': '15|13|13|10|9|9|11|12|10|11|12|14|14|15|0'}
+
+For the `dictionary` configuration:
+{'label': 0.7361099720001221, 
+'phrase': 'still charming'}
+
+For the `ptb` configuration:
+{'ptb_tree': '(3 (2 Yet) (3 (2 (2 the) (2 act)) (3 (4 (3 (2 is) (3 (2 still) (4 charming))) (2 here)) (2 .))))'}
```
Comment (lhoestq) [inline datasets/sst/README.md:92]:
````suggestion
For the `default` configuration:
```
{'label': 0.7222200036048889,
 'sentence': 'Yet the act is still charming here .',
 'tokens': 'Yet|the|act|is|still|charming|here|.',
 'tree': '15|13|13|10|9|9|11|12|10|11|12|14|14|15|0'}
```

For the `dictionary` configuration:
```
{'label': 0.7361099720001221, 
'phrase': 'still charming'}
```

For the `ptb` configuration:
```
{'ptb_tree': '(3 (2 Yet) (3 (2 (2 the) (2 act)) (3 (4 (3 (2 is) (3 (2 still) (4 charming))) (2 here)) (2 .))))'}
```
````

File: datasets/sst/sst.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+_DESCRIPTION = """\
+The Stanford Sentiment Treebank, the first corpus with fully labeled parse trees that allows for a
+complete analysis of the compositional effects of sentiment in language.
+"""
+
+_HOMEPAGE = "https://nlp.stanford.edu/sentiment/"
+
+_LICENSE = ""
+
+_URL = "https://www.dropbox.com/s/9j9dc55hs28wrye/stanfordSentimentTreebank.zip?dl=1"
```
Comment (lhoestq) [inline datasets/sst/sst.py:49]: This is not the official download link right ?
Why not use the official link ?

File: datasets/sst/sst.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            supervised_keys=None,
+            homepage=_HOMEPAGE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        data_dir = dl_manager.download_and_extract(_URL)
+
+        file_paths = {}
+        for split_index in range(0, 4):
```
Comment (lhoestq) [inline datasets/sst/sst.py:107]: Why do you need to copy the dictionary 4 times ?

**Review by lhoestq — COMMENTED** [review]
Awesome thank you :)

I added some suggestions and I also have a question regarding the download link:

File: datasets/sst/sst.py (hunk 3 above)
Comment (patpizio) [reply datasets/sst/sst.py:49]: There is an [issue](https://groups.google.com/g/word2vec-toolkit/c/QIUjLw6RqFk/m/_iEeyt428wkJ) with the official files: `datasetSentences.txt` is almost impossible to decode. Every "special" symbol like â or ç results in an unreadable sentence whenever I look for its label in `dictionary.txt` and `sentiment_labels.txt`. The only solution I found was to manually replace all the special characters and save the resulting file in `utf-8`. 
Of course this is a bit patchy and I'd be happy for the data to be hosted on a more "official" server or, even better, find what encoding was used in the original file (which I kept in the repo as `datasetSentences_old_encoding.txt`. (You're welcome to give it a shot) 😜

Comment (patpizio) [reply datasets/sst/sst.py:49]: I think I tried almost all of the [Standard Encodings](https://docs.python.org/3/library/codecs.html#standard-encodings)

File: datasets/sst/sst.py (hunk 4 above)
Comment (patpizio) [reply datasets/sst/sst.py:107]: 3 for train, validation and test in the `default` configuration. The last one is for the only split in the `dictionary` configuration. I am not sure if there's a more clever way to do this

File: datasets/sst/sst.py (hunk 3 above)
Comment (lhoestq) [reply datasets/sst/sst.py:49]: I managed to read the old encoding this way:
```python
datasetSentences = open("datasetSentences.txt", encoding="utf-8")
datasetSentences_old_encoding = open("datasetSentences_old_encoding.txt", encoding="utf-8")

for a, b in zip(datasetSentences, datasetSentences_old_encoding):
    b = b.encode("utf-8").replace(b"\xc3\x83\xc2", b"\xc3").replace(b"\xc3\x82\xc2", b"\xc2").decode("utf-8")
    b = b.replace("-LRB-", "(").replace("-RRB-", ")")
    assert a == b
```
Using these two `replace` calls, you can read the original `datasetSentences.txt` file.

Comment (patpizio) [reply datasets/sst/sst.py:49]: 🤦‍♂️ Of course! It had to be `\xc3\x83\xc2`! 🤣
Whenever you have time, let me know how did you get to that solution...

I'll go on and edit my code accordingly, thanks a lot for the fix 👍

Comment (lhoestq) [reply datasets/sst/sst.py:49]: Haha yea that's really weird x)

I just compared the binary string of the the first badly encoded sentence with the corrected sentence and noticed that the only difference was these bad characters: `\x83\xc2` after `\xc3`. By chance removing these characters fixed all the encoding issues at once.

Comment (patpizio) [reply datasets/sst/sst.py:49]: Brilliant!

File: datasets/sst/sst.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+_DESCRIPTION = """\
+The Stanford Sentiment Treebank, the first corpus with fully labeled parse trees that allows for a
+complete analysis of the compositional effects of sentiment in language.
+"""
+
+_HOMEPAGE = "https://nlp.stanford.edu/sentiment/"
+
+_LICENSE = ""
+
+_URL = "https://www.dropbox.com/s/9j9dc55hs28wrye/stanfordSentimentTreebank.zip?dl=1"
```
Comment (lhoestq) [inline datasets/sst/sst.py:49]:
```suggestion
```
we don't need this anymore ;)

File: datasets/sst/sst.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                for i, row in enumerate(tree_reader, start=1):
+                    trees[i]["tree"] = row[0]
+
+            with open(splits_path, encoding="utf-8") as spl, open(sentences_path, encoding="utf-8") as snt:
+                splits_reader = csv.DictReader(spl, delimiter=",", quoting=csv.QUOTE_NONE)
+                splits = {row["sentence_index"]: row["splitset_label"] for row in splits_reader}
+
+                sentence_reader = csv.DictReader(snt, delimiter="\t", quoting=csv.QUOTE_NONE)
+                for id_, row in enumerate(sentence_reader):
+                    row["sentence"] = (
```
Comment (lhoestq) [inline datasets/sst/sst.py:193]:
```suggestion
                    # fix encoding, see https://github.com/huggingface/datasets/pull/1961#discussion_r585969890
                    row["sentence"] = (
```

**Review by lhoestq — APPROVED** [review]
Thanks for the changes !
LGTM :)

**Pull request merged**